strawberryfields/apps/graph/dense.py

```diff
@@ -0,0 +1,179 @@
+        nodes (int): the size of desired dense subgraph
+        iterations (int): number of iterations to use in algorithm
+        options (dict[str, dict[str, Any]]): dict-of-dicts specifying options in different parts
+            of heuristic search algorithm; defaults to :const:`OPTIONS_DEFAULTS`
+
+    Returns:
+        Tuple[float, list]: the density and list of nodes corresponding to the densest subgraph
+        found
+    """
+    options = {**OPTIONS_DEFAULTS, **(options or {})}
+
+    method = options["heuristic"]["method"]
+
+    if not callable(method):
+        method = METHOD_DICT[method]
+
+    return method(
+        graph=utils.to_networkx_graph(graph),
+        nodes=nodes,
+        iterations=iterations,
+        options=options,
+    )
+
+
+def random_search(
+    graph: nx.Graph, nodes: int, iterations: int = 1, options: Optional[dict] = None
+) -> Tuple[float, list]:
+    """Random search algorithm for finding dense subgraphs of a given size.
+
+    The algorithm proceeds by sampling subgraphs according to the
+    :func:`~strawberryfields.future.apps.graph.sample.sample_subgraphs`. The resultant subgraphs
+    are resized using :func:`~strawberryfields.future.apps.graph.resize.resize_subgraphs` to
+    be of size ``nodes``. The densest subgraph is then selected among all the resultant
+    subgraphs. Specified``options`` must be of the form given in :func:`find_dense`.
+
+    Args:
+        graph (nx.Graph): the input graph
+        nodes (int): the size of desired dense subgraph
+        iterations (int): number of iterations to use in algorithm
+        options (dict[str, dict[str, Any]]): dict-of-dicts specifying options in different parts
+            of heuristic search algorithm; defaults to :const:`OPTIONS_DEFAULTS`
+
+    Returns:
+        Tuple[float, list]: the density and list of nodes corresponding to the densest subgraph
+        found
+    """
+    options = {**OPTIONS_DEFAULTS, **(options or {})}
+
+    samples = sample.sample_subgraphs(
+        graph=graph,
+        nodes=nodes,
+        samples=iterations,
+        sample_options=options["sample"],
+        backend_options=options["backend"],
+    )
+
+    samples = resize.resize_subgraphs(
+        subgraphs=samples, graph=graph, target=nodes, resize_options=options["resize"]
+    )
+
+    density_and_samples = [(nx.density(graph.subgraph(s)), s) for s in samples]
+
+    return max(density_and_samples)
+
+
+METHOD_DICT = {"random-search": random_search}
+"""Dict[str, func]: Included methods for finding dense subgraphs. The dictionary keys are strings
+describing the method, while the dictionary values are callable functions corresponding to the
+method."""
+
+OPTIONS_DEFAULTS = {
+    "heuristic": {"method": random_search},
+    "backend": BACKEND_DEFAULTS,
+    "resize": resize.RESIZE_DEFAULTS,
+    "sample": sample.SAMPLE_DEFAULTS,
+}
+"""Dict[str, dict[str, Any]]: Options for dense subgraph identification heuristics. Composed of a
```

Comment:
```suggestion
"""Dict[str, Dict[str, Any]]: Options for dense subgraph identification heuristics. Composed of a
```

Reply:
Thanks @smite, for now I went with the all lowercase version (`dict[str, dict[str, Any]]`) but happy to open up the conversation for which to use as it seems there's arguments for both.

Reply:
@trbromley @smite after reading through all the comments I'm happy with either standard. I think the main takeaway is consistency though, we should try and stick to the same style through the entire codebase